File: Data Structures & Algorithms/spiral-matrix/submission-4.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int n = matrix.size()*matrix[0].size();
        for(int i=0;i<=(int)max(matrix.size(), matrix[0].size())/2;i++) {
            for(int j=i;res.size()<n&&j<(int)matrix[0].size()-i-1;j++) {
                res.push_back(matrix[i][j]);
            }
            for(int j=i;res.size()<n&&j<(int)matrix.size()-i-1;j++) {
                res.push_back(matrix[j][matrix[0].size()-1-i]);
            }
            for(int j=(int)matrix[0].size()-i-1;res.size()<n&&j>i;j--) {
                res.push_back(matrix[matrix.size()-i-1][j]);
            }
            for(int j=(int)matrix.size()-i-1;res.size()<n&&j>i;j--) {
                res.push_back(matrix[j][i]);
            }
        }
        if(res.size()<n) res.push_back(matrix[matrix.size()/2][matrix.size()/2]);
        return res;
    }
};

File: Data Structures & Algorithms/single-threaded-cpu/submission-4.java
class Solution {
    public int[] getOrder(int[][] tasks) {
        int[] index = IntStream.range(0, tasks.length)
            .boxed()
            .sorted((a, b)-> {
                if(tasks[a][0]==tasks[b][0]) {
                    return tasks[a][1]-tasks[b][1];
                }
                return tasks[a][0]-tasks[b][0];
            })
            .mapToInt(a->a)
            .toArray();

        Queue<Integer> pq = new PriorityQueue<>((a, b) -> tasks[a][1]-tasks[b][1]);
        List<Integer> res = new ArrayList<>();
        pq.add(index[0]);
        int i=1;
        int bound = 0;
        while(res.size()<index.length) {
            int nxt = pq.remove();
            res.add(nxt);
            bound = Math.max(bound, tasks[nxt][0])+tasks[nxt][1];
            while(i<tasks.length && tasks[index[i]][0]<=bound) {
                pq.add(index[i]);
                i++;
            }
            if(pq.isEmpty() && i<tasks.length) pq.add(index[i++]);
        }
        return res.stream()
            .mapToInt(e->e)
            .toArray();
    }
}